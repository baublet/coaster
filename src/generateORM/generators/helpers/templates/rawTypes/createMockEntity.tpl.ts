function $$functionName(
  defaults: Partial<$$prefixedEntityName> = {}
): $$prefixedEntityName {
  const test$$prefixedEntityName: $$prefixedEntityName = {
    $$columnDefaults,
    ...defaults,
  };
  return test$$prefixedEntityName;
}
