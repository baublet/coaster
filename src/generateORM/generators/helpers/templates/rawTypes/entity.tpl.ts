$$comment
export $$interfaceOrType $$entityName $$typeEqualsSign{
$$columns
};
