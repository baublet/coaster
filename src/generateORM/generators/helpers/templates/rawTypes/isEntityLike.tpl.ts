export function $$functionName(subject: any): subject is $$prefixedEntityName {
  if (typeof subject === "object") {
    if (objectHasProperties(subject, $$columnNamesAsJsonString)) {
      return true;
    }
  }
  return false;
}
