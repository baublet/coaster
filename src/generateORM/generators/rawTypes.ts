import { MetaData } from ".";
import { RawSchema } from "../drivers";
import { getName } from "./helpers";

export const rawTypes = (
  schema: RawSchema,
  metaData: MetaData,
  options: {
    typesOrInterfaces: "types" | "interfaces";
    prefixSchemaName?: boolean;
    rawPrefix?: string;
    getTypeName?: (
      schema: string,
      table: string,
      column: string
    ) => string | undefined;
  } = {
    typesOrInterfaces: "interfaces",
    prefixSchemaName: false,
    getTypeName: () => undefined,
  }
) => {
  let code = "";
  const rawPrefix = options.rawPrefix === undefined ? "Raw" : options.rawPrefix;

  for (const table of schema.tables) {
    const tableName = getName(
      schema.name,
      table.name,
      undefined,
      options.prefixSchemaName
    );

    const schemaAndTablePath = `${schema.name ? schema.name + "." : ""}${
      table.name
    }`;
    metaData.tableEntityNames.set(schemaAndTablePath, rawPrefix + tableName);
    metaData.entityTableNames.set(rawPrefix + tableName, schemaAndTablePath);

    if (table.comment) {
      code += `\n/** ${table.comment} */\n`;
    }

    code += `${
      options.typesOrInterfaces === "interfaces" ? "interface" : "type"
    } ${rawPrefix}${tableName} ${
      options.typesOrInterfaces === "interfaces" ? "" : "= "
    }{`;

    for (const column of table.columns) {
      if (column.comment) {
        code += `\n/** ${column.comment} */`;
      }
      code += `\n${column.name}`;
      code += column.nullable ? "?: " : ": ";
      code +=
        options.getTypeName?.(schema.name, table.name, column.name) ||
        column.type;
      code += ";";
    }

    code += `\n};\n`;
  }

  return code;
};
