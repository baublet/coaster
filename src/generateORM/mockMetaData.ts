import { MetaData } from "./generators";
import { getTemplateManager } from "../generateORM/generators/helpers";

export function getMockMetaData(): MetaData {
  return {
    testConnectionVariable: "connection",
    codeOutputFullPath: "./generated.ts",
    generateTestCode: false,
    entityTableNames: new Map(),
    setHeader: jest.fn(),
    setTestHeader: jest.fn(),
    tableEntityNames: new Map(),
    rawBaseQueryFunctionNames: new Map(),
    transformerFunctionNames: {},
    typeAssertionFunctionNames: new Map(),
    typeGuardFunctionNames: new Map(),
    tableRawEntityNames: new Map(),
    namedEntityColumnNames: new Map(),
    namedEntityInputTypeNames: new Map(),
    namedDatabaseEnumNames: new Map(),
    rawDatabaseEnumNames: new Map(),
    namedCreateTestEntityFunctionNames: new Map(),
    rawCreateTestEntityFunctionNames: new Map(),
    rawEnumValues: new Map(),
    templateManager: getTemplateManager(),
    namedMockEntityFunctionNames: new Map(),
    rawMockEntityFunctionNames: new Map(),
  };
}
