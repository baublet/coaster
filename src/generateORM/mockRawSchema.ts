import { RawSchema } from "./drivers";

export const getMockRawSchema: () => RawSchema = () => ({
  name: "public",
  flavor: "pg",
  enums: [
    {
      name: "user_account_source",
      values: ["github", "bitbucket"],
    },
  ],
  tables: [
    {
      name: "user",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "string",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "name",
          comment: "Users don't have to provide their name",
          columnType: "columnType",
          nullable: true,
          type: "string",
          foreignKeys: [],
          hasDefault: false,
        },
        {
          name: "data",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
    {
      name: "user_account",
      comment: "Singular users can have multiple accounts",
      primaryKeyColumn: "id",
      primaryKeyType: "string",
      uniqueConstraints: [],
      columns: [
        {
          name: "id",
          columnType: "columnType",
          nullable: false,
          type: "unknown",
          foreignKeys: [],
          hasDefault: true,
        },
        {
          name: "source",
          comment: "Who sent us this users' data?",
          columnType: "columnType",
          nullable: false,
          type: "enum",
          foreignKeys: [],
          hasDefault: false,
          enumPath: "public.user_account_source",
        },
        {
          name: "source_data",
          comment: "The raw data sent from the source provider",
          columnType: "columnType",
          nullable: false,
          type: "AnyJson",
          foreignKeys: [],
          hasDefault: false,
        },
      ],
    },
  ],
});
