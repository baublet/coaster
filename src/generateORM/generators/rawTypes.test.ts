import { rawTypes } from "./rawTypes";
import { mockRawSchema } from "../mockRawSchema";
import { getMockMetaData } from "../mockMetaData";

describe("interfaces/types", () => {
  it("spits out types", () => {
    expect(
      rawTypes(mockRawSchema, getMockMetaData(), { typesOrInterfaces: "types" })
    ).toMatchSnapshot();
  });

  it("spits out interfaces", () => {
    expect(
      rawTypes(mockRawSchema, getMockMetaData(), {
        typesOrInterfaces: "interfaces",
      })
    ).toMatchSnapshot();
  });
});
