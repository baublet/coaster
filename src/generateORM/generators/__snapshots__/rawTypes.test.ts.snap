// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interfaces/types spits out interfaces 1`] = `
"interface RawUser {
id: unknown;
/** Users don't have to provide their name */
name?: string;
data: JSON;
};

/** Singular users can have multiple accounts */
interface RawUserAccount {
id: unknown;
/** Who sent us this users' data? */
source: string;
/** The raw data sent from the source provider */
source_data: JSON;
};
"
`;

exports[`interfaces/types spits out types 1`] = `
"type RawUser = {
id: unknown;
/** Users don't have to provide their name */
name?: string;
data: JSON;
};

/** Singular users can have multiple accounts */
type RawUserAccount = {
id: unknown;
/** Who sent us this users' data? */
source: string;
/** The raw data sent from the source provider */
source_data: JSON;
};
"
`;
